ext/semian_cb_data/semian_cb_data.c

```diff
@@ -0,0 +1,900 @@
+    rb_raise(rb_eTypeError, "expected integer for arr_max_size");
+  if (TYPE(permissions) != T_FIXNUM)
+    rb_raise(rb_eTypeError, "expected integer for permissions");
+
+  if (NUM2SIZET(size) <= 0)
+    rb_raise(rb_eArgError, "arr_max_size must be larger than 0");
+
+  const char *id_str = NULL;
+  if (TYPE(id) == T_SYMBOL) {
+    id_str = rb_id2name(rb_to_id(id));
+  } else if (TYPE(id) == T_STRING) {
+    id_str = RSTRING_PTR(id);
+  }
+  ptr->key = generate_key(id_str);
+  //rb_warn("converted name %s to key %d", id_str, ptr->key);
+
+  // Guarantee arr_max_size >=1 or error thrown
+  ptr->arr_max_size = NUM2SIZET(size);
+
+  // id's default to -1
+  ptr->semid = -1;
+  ptr->shmid = -1;
+  // addresses default to NULL
+  ptr->shm_address = 0;
+  ptr->lock_triggered = false;
+
+  semian_cb_data_acquire_semaphore(self, permissions);
+  semian_cb_data_acquire_memory(self, permissions);
+
+  return self;
+}
+
+static VALUE
+semian_cb_data_clean(VALUE self)
+{
+  semian_cb_data_delete_memory(self);
+  semian_cb_data_delete_semaphore(self);
+  return self;
+}
+
+
+
+/*
+ * Functions set_semaphore_permissions, configure_tickets, create_semaphore
+ * are taken from semian.c with extra code removed
+ */
+static void
+set_semaphore_permissions(int sem_id, int permissions)
+{
+  union semun sem_opts;
+  struct semid_ds stat_buf;
+
+  sem_opts.buf = &stat_buf;
+  semctl(sem_id, 0, IPC_STAT, sem_opts);
+  if ((stat_buf.sem_perm.mode & 0xfff) != permissions) {
+    stat_buf.sem_perm.mode &= ~0xfff;
+    stat_buf.sem_perm.mode |= permissions;
+    semctl(sem_id, 0, IPC_SET, sem_opts);
+  }
+}
+
+
+static void
+configure_tickets(int sem_id, int tickets, int should_initialize)
+{
+  struct timeval start_time, cur_time;
+
+  if (should_initialize) {
+    if (-1 == semctl(sem_id, 0, SETVAL, kCBTicketMax)) {
+      rb_warn("semctl: failed to set semaphore with semid %d, position 0 to %d", sem_id, 1);
+      raise_semian_syscall_error("semctl()", errno);
+    } else {
+      //rb_warn("semctl: set semaphore with semid %d, position 0 to %d", sem_id, 1);
+    }
+  } else if (tickets > 0) {
+    // it's possible that we haven't actually initialized the
+    // semaphore structure yet - wait a bit in that case
+    int ret;
+    if (0 == (ret = semctl(sem_id, 0, GETVAL))) {
+      gettimeofday(&start_time, NULL);
+      while (0 == (ret = semctl(sem_id, 0, GETVAL))) { // loop while value == 0
+        usleep(10000); /* 10ms */
+        gettimeofday(&cur_time, NULL);
+        if ((cur_time.tv_sec - start_time.tv_sec) > kCBInitializeWaitTimeout) {
+          rb_raise(eInternal, "timeout waiting for semaphore initialization");
+        }
+      }
+      if (-1 == ret) {
+        rb_raise(eInternal, "error getting max ticket count, errno: %d (%s)", errno, strerror(errno));
+      }
+    }
+
+    // Rest of the function (originally from semian.c) was removed since it isn't needed
+  }
+}
+
+static int
+create_semaphore(int key, int permissions, int *created)
```

Comment:
Should the constants be reused as well? Or abstracting that and passing that in as arguments?